modules/packages/LinearAlgebra.chpl

```diff
@@ -0,0 +1,733 @@
+  return A;
+}
+
+
+/* Return an identity matrix over domain ``Dom`` */
+proc eye(Dom: domain(2), type eltType=real) {
+  var A: [Dom] eltType;
+  const diagonal = min(Dom.shape(1), Dom.shape(2));
+  return eye(Dom.shape(1), Dom.shape(2), eltType=eltType);
+}
+
+
+//
+// Matrix Operations
+//
+
+pragma "no doc"
+inline proc transpose(D: domain(1)) {
+  return D;
+}
+
+inline proc transpose(D: domain(2)) {
+  return {D.dim(2), D.dim(1)};
+}
+
+
+pragma "no doc"
+/* Transpose vector (no-op) */
+inline proc transpose(A: [?Dom]) where Dom.rank == 1 {
+  return A;
+}
+
+
+pragma "no doc"
+/* Syntactic sugar for transpose(Vector) */
+proc _array.T where this.domain.rank == 1 { return transpose(this); }
+
+
+/* Transpose vector or matrix.
+
+   .. note::
+
+      Since row vectors and columns vectors are indistinguishable, passing
+      a vector to this function will return that vector unchanged
+
+*/
+proc transpose(A: [?Dom] ?eltType) where Dom.rank == 2 {
+  if !isBLASType(eltType) then
+    return _transpose(A);
+  else if Dom.shape(1) == 1 then
+    return reshape(A, transpose(Dom));
+  else if Dom.shape(2) == 1 then
+    return reshape(A, transpose(Dom));
+  else {
+    var B: [Dom] eltType = eye(Dom, eltType=eltType);
+    var C: [transpose(Dom)] eltType;
+    // TODO -- use native algorithm (BLAS not necessary or memory-efficient)
+    gemm(A, B, C, 1:eltType, 0:eltType, opA=Op.T);
+    return C;
+  }
+}
+
+
+/* Transpose vector or matrix */
+proc _array.T where this.domain.rank == 2 { return transpose(this); }
+
+
+/* Transpose matrix with native implementation */
+pragma "no doc"
+private proc _transpose(A: [?Dom] ?eltType) {
+  var C: [transpose(Dom)] eltType;
+
+  // naive algorithm
+  forall (i, j) in Dom do
+    C[j, i] = A[i, j];
+
+  return C;
+}
+
+
+/* Add matrices, maintaining dimensions */
+proc matPlus(A: [?Adom] ?eltType, B: [?Bdom] eltType) {
+  if Adom.rank != Bdom.rank then compilerError("Ranks to not match");
+  if Adom.size != Bdom.size then halt("Unmatched array size");
+  var C: [Adom] eltType = A + B;
+  return C;
+}
+
+
+/* Subtract matrices, maintaining dimensions */
+proc matMinus(A: [?Adom] ?eltType, B: [?Bdom] eltType) {
+  if Adom.rank != Bdom.rank then compilerError("Ranks to not match");
+  if Adom.size != Bdom.size then halt("Unmatched array size");
+  var C: [Adom] eltType = A - B;
+  return C;
+}
+
+/*
+    Generic matrix multiplication, ``A`` and ``B`` can be a matrix, vector, or
+    scalar
+*/
+proc dot(A: [?Adom] ?eltType, B: [?Bdom] eltType) {
+  // vector-vector
+  if Adom.rank == 1 && Bdom.rank == 1 then
+    return inner(A, B);
+  // matrix-(vector|matrix)
+  else
+    return matMult(A, B);
+}
+
+
+pragma "no doc"
+/* Element-wise scalar multiplication */
+proc dot(A: [?Adom] ?eltType, b) where isNumeric(b) {
+  var C: [Adom] eltType = A*b;
+  return C;
+}
+
+
+pragma "no doc"
+/* Element-wise scalar multiplication */
+proc dot(a, B: []) where isNumeric(a) {
+  return dot(B, a);
+}
+
+
```

Comment:
Hi @npadmana -- I'm happy that you commented on this:

To clarify, I *intend* for `dot` to be the general matrix-matrix multiplication routine, which  wraps `inner` and `matMult`. We leave `inner` available as part of the public interface in case users want to explicitly distinguish `inner` vs `outer` products of vectors (since row vs. column vector is indistinguishable.)

I also left `matMult` as part of the public interface trying to mirror what I did with `inner`, but it's not really useful aside from being an alternative syntax. As you point out, it's a potential source of confusion, so I'm inclined to make it private for this first version of the library.